File: nova_example/quarter_truck/nova_quarter_truck_ssp.py

import os
import sys
from pathlib import Path

# Add project root to sys.path to import nova_sim_py
project_root = Path(__file__).parent.parent.parent
sys.path.append(str(project_root))

from nova_sim_py import NovaSimulation, NovaSimulationStructure, EcosLib

def main():
    print(f"Nova Ecoslib version: {EcosLib.version()}")

    EcosLib.set_log_level("debug")

    ssp_dir = (project_root / 'data' / 'ssp' / '1.0' / 'quarter_truck').resolve()
    
    result_dir = project_root / 'results' / 'python'
    result_dir.mkdir(parents=True, exist_ok=True)
    result_file = str(result_dir / "nova_quarter_truck_ssp_py.csv")

    with NovaSimulation(ssp_path=str(ssp_dir), step_size=1.0 / 100) as sim:
        sim.add_csv_writer(result_file, str(ssp_dir / "CsvConfig.xml"))

        sim.init(parameter_set="initialValues")
        sim.step_until(10)
        sim.terminate()

    print(f"Nova Python SSP Quarter-Truck finished. Results: {result_file}")

if __name__ == "__main__":
    main()

File: nova_example/quarter_truck/nova_quarter_truck_ssp.cpp

#include "nova/components/algorithm/fixed_step_algorithm.hpp"
#include "nova/components/recorder/csv_recorder.hpp"
#include "nova/components/logger/logger.hpp"
#include "nova/components/ssp/ssp_loader.hpp"
#include "nova/components/util/chart_plotter.hpp"

using namespace nova_sim;

int main()
{
    set_logging_level(log::level::debug);

    const std::filesystem::path sspFolder = std::string(DATA_FOLDER) + "/ssp/1.0/quarter_truck";

    try {
        const auto ss = load_ssp(sspFolder);
        const auto sim = ss->load(std::make_unique<fixed_step_algorithm>(1.0 / 100));

        std::filesystem::create_directories(RESULT_FOLDER);
        auto csvWriter = std::make_unique<csv_recorder>(std::string(RESULT_FOLDER) + "/nova_quarter_truck_ssp.csv");
        const auto outputPath = csvWriter->output_path();
        
        csv_config& config = csvWriter->config();
        config.register_variable("chassis::zChassis");
        config.register_variable("wheel::zWheel");
        config.register_variable("ground::zGround");
        sim->add_listener("csv_recorder", std::move(csvWriter));

        log::Stopwatch sw;
        sim->init("initialValues");
        sim->step_until(5);
        log::info("Nova SSP Quarter-Truck finished. Elapsed {}s", sw.elapsed().count());

        sim->terminate();

        plot_csv(outputPath, sspFolder / "ChartConfig.xml");
    } catch (const std::exception& ex) {

        log::err(ex.what());
    }
}
